src/librbd/io/ReadResult.cc

```diff
@@ -141,14 +141,18 @@ void ReadResult::C_ObjectReadRequest::finish(int r) {
     aio_completion->lock.lock();
     if (!extent_map.empty()) {
       aio_completion->read_result.m_destriper.add_partial_sparse_result(
-        cct, bl, extent_map, object_off, buffer_extents);
+        cct, bl, extent_map, object_extents, buffer_extents);
     } else {
       // handle the case where a sparse-read wasn't issued
       aio_completion->read_result.m_destriper.add_partial_result(
```

Comment:
Nit: we need to handle this path as well since librbd will automatically use non-sparse reads for small IOs. i.e. I would expect one buffer extent for each object extent since it might not be linearly mapped.

Reply:
Actually, it seems to me that following my previous PR adding reading multiple extents, extent_map here will never be empty, and thus the else clause above is dead code.
Can you verify?

Reply:
Specifically:
https://github.com/ceph/ceph/blob/148c18c6d621b6053e8c2d56d65051447a41f7ab/src/librbd/io/ObjectRequest.cc#L260

Reply:
Indeed. Can you yank the dead code path as part of this PR?

Reply:
I think I was not totally correct.
It seems that if the request was served by the CORE layer it is guaranteed that extent_map won't be empty.
However, if it came from another layer, say CACHE, maybe it can still be empty?
Even if it can be empty, this is not relevant for the multi-extent case I'm addressing here, since multi-extent requests will always go down to the core layer.

Reply:
Actually, now that I think about it, there are older OSDs that will happily return an empty extent map for sparse reads so historically this code in `ReadResult` used to check if `extent_map` was empty. If so, it applied the logic you added `ObjectRequest` for inserting the offset and length. I suppose that could be re-added to `ReadResult` so that there isn't the possibility (now or in the future) to get into a case where multiple object extents were issued.

Reply:
Sorry I don't follow.
You want to protect against a (possibly future) case where the request is for multiple extents, yet extent_map is empty?
In that case, how can we tell to which extent does `bl` belongs to?

Reply:
One possibility would be to swap out the `bufferlist` and `Extents` parameters to `ObjectReadRequest` and instead pass a `std::vector<std::pair<bufferlist, Extents>>` typedef-ed parameter so that `C_ObjectReadRequest` would own that data structure and could cleanly re-assemble the results.

Reply:
Today we have in ObjectRequest:
```  typedef std::pair<ceph::bufferlist, Extents> ExtentResult;
  typedef std::vector<ExtentResult> ExtentResults;
  ExtentResults m_extent_results;```
So you want to move this to C_ObjectReadRequest?
And then, instead of merging ExtentResults to a single sparse bufferlist as we do today, we should iterate them and call the striper.add_partial{_sparse}_result?

Reply:
I would put the types in `io/Types.h` but otherwise that sounds good to me. Then we have a single code path for handling the re-assembly of read requests regardless of the layer.

Reply:
And changing ObjectDispatchInterface.read to get `*ExtentResults`, instead of (`ceph::bufferlist* , Extents*`). Right?
And the assumption is that the order is ExtentResults should match the order of the input extents `const Extents &extents`.

Reply:
> And changing ObjectDispatchInterface.read to get *ExtentResults, instead of (ceph::bufferlist* , Extents*). Right?

Right.

>  And the assumption is that the order is ExtentResults should match the order of the input extents const Extents &extents.

Right -- or tweak `ExtentResults` to be an `(object_offset, bufferlist, Extents)` tuple/struct to keep it explicit.

Reply:
OK. My current plan is to define:
```
typedef std::pair<ceph::bufferlist, Extents> ReadResult;
typedef std::pair<Extent, ReadResult> ReadExtent;
typedef std::vector<ReadExtent> ReadExtents;
```

And have ObjectDispatchInterface.read get *ReadExtents.

Since ReadResult is a pair of actual bufferlist and Extents objects, I'm adding address taking operator in all usages to avoid copying these objects.

For example, in ObjectCacherObjectDispatch:
```
  auto read_extent = &(read_extents->front());
  auto [object_off, object_len] = read_extent->first;
  auto read_result = &(read_extent.second);
```

Otherwise, if I define ReadResult as a pair of pointers. I would need to explicitly allocate the bufferlists and extents.

Thoughts?

Reply:
Or actually is it ok to just copy the std::pair, as it is not doing a deep copy resulting in copying the bufferlist and extents nested inside?

Reply:
You can use `auto&` to get a reference to the members of the pair. Will you need the object length portion of the object extent, though?

Reply:
changes pushed + rebase